test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs

```diff
@@ -0,0 +1,211 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.TestHelpers;
+
+namespace Datadog.Trace.ClrProfiler.IntegrationTests
+{
+    /// <summary>
+    /// Base class for all span expectations. Inherit from this class to extend per type or integration.
+    /// </summary>
+    public class SpanExpectation
+    {
+        public SpanExpectation(string serviceName, string operationName, string type)
+        {
+            ServiceName = serviceName;
+            OperationName = operationName;
+            Type = type;
+
+            // Expectations for all spans regardless of type should go here
+            RegisterDelegateExpectation(ExpectBasicSpanDataExists);
+
+            RegisterCustomExpectation(nameof(OperationName), actual: s => s.Name, expected => OperationName);
+            RegisterCustomExpectation(nameof(ServiceName), actual: s => s.Service, expected => ServiceName);
+            RegisterCustomExpectation(nameof(Type), actual: s => s.Type, expected => Type);
+            RegisterCustomExpectation(nameof(ResourceName), actual: s => s.Resource.TrimEnd(), expected => ResourceName);
+
+            RegisterTagExpectation(
```

Comment:
This seems to be the only `RegisterTagExpectation` that uses the `when` parameter. What are your thoughts on changing this semantic check to using the `RegisterDelegateExpectations` method instead and simplifying `RegisterTagExpectation` to not have the additional `when` parameter?

Reply:
I'm fine with that, but I do like it being explicit. I think it forces the person writing the test to think about *when* tags apply and when they don't. Thoughts?

Reply:
Yeah that's a good point. Can either have two overloads or make `when` a default argument that defaults to `Always`? That way we don't have to specify it everywhere

Reply:
sure